r1cs-std/src/groups/curves/twisted_edwards/mod.rs

```diff
@@ -757,6 +957,122 @@ mod projective_impl {
+                    .add_bool_with_coeff(cs.ns(|| format!("add bool 01 in window {}", i)), &bits[0], x_coeffs[1] - &x_coeffs[0])?
+                    .add_bool_with_coeff(cs.ns(|| format!("add bool 10 in window {}", i)), &bits[1], x_coeffs[2] - &x_coeffs[0])?
+                    .add_bool_with_coeff(cs.ns(|| format!("add bool 11 in window {}", i)), &precomp, x_coeffs[3] - &x_coeffs[2] - &x_coeffs[1] + &x_coeffs[0])?;
+
+
+                let y = F::three_bit_cond_neg_lookup(
+                    cs.ns(|| format!("y lookup in window {}", i)),
+                    &bits,
+                    &precomp,
+                    &y_coeffs,
+                )?;
+
+                let tmp = MontgomeryAffineGadget::new(x, y);
+
+                match result {
+                    None => {
+                        result = Some(tmp);
+                    }
+                    Some(ref mut result) => {
+                        *result = tmp.add(
+                            cs.ns(|| format!("addition of window {}", i)),
+                            result,
+                        )?;
+                    }
+                }
+
+                if i > 0 && (i + 1) % segment_size == 0 {
+                    let segment_result = result.unwrap();
+                    let segment_result = segment_result.into_edwards(
+                        cs.ns(|| format!("segment result in window {}", i)),
+                    )?;
+                    match edwards_result {
+                        None => {
+                            edwards_result = Some(segment_result);
+                        }
+                        Some(ref mut edwards_result) => {
+                            *edwards_result = GroupGadget::<TEAffine<P>, ConstraintF>::add(
+                                &segment_result,
+                                cs.ns(|| format!("edwards addition of window {}", i)),
+                                edwards_result,
+                            )?;
+                        }
+                    }
+                    result = None;
+                }
+            }
+            if result.is_some() {
+                let segment_result = result.unwrap();
+                let segment_result = segment_result.into_edwards(
+                    cs.ns(|| "segment result in leftover"),
+                )?;
+                match edwards_result {
+                    None => {
+                        edwards_result = Some(segment_result);
+                    }
+                    Some(ref mut edwards_result) => {
+                        *edwards_result = GroupGadget::<TEAffine<P>, ConstraintF>::add(
+                            &segment_result,
+                            cs.ns(|| "edwards addition of leftover"),
+                            edwards_result,
+                        )?;
+                    }
+                }
+            }
```

Comment:
These pieces of code are almost exactly the same, so it might be good to extract it into a macro or function/closure that is local to `precomputed_base_scalar_mul_3_bit_with_conditional_negation`?

Reply:
Take a look at the latest commit